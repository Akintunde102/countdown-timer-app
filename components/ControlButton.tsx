import React, {ReactChild} from 'react';
import {
  StyleSheet,
  Text,
  TouchableOpacity,
  ViewStyle,
  View,
} from 'react-native';

const ControlButton = ({
  action,
  customStyle,
  children,
}: {
  action: Function;
  customStyle: ViewStyle;
  children: ReactChild;
}) => {
  return (
    <TouchableOpacity
      style={{...customStyle}}
      onPress={() => {
        action();
      }}>
      {children}
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  view: {},
});

export default ControlButton;
