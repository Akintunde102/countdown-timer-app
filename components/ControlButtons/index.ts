import ControlButtons from './ControlButtons';

export default ControlButtons;
