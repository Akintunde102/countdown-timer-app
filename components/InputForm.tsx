import React, {useState} from 'react';
import {
  StyleSheet,
  View,
  TextInput,
  Text,
  TouchableOpacity,
  ViewStyle,
} from 'react-native';

const InputForm = ({
  setDuration,
  setTimerStatus,
  style
}: {
  setDuration: Function;
  setTimerStatus: Function;
  style: ViewStyle
}) => {
  const [durationInput, setDurationInput] = useState<number>(0);
  return (
    <View style={{...style, ...styles.view}}>
      <View style={styles.textView}>
        <Text style={styles.text}>CountDown:</Text>
      </View>
      <View style={styles.inputView}>
        <TextInput
          style={styles.input}
          placeholder="(Min)"
          placeholderTextColor="#777776"
          onChangeText={(text: string) => {
            setDurationInput(parseInt(text, 10));
          }}
        />
      </View>
      <TouchableOpacity
        style={styles.button}
        onPress={() => {
          setDuration(durationInput);
          setTimerStatus({starts: true});
        }}>
        <Text style={styles.buttonText}>START</Text>
      </TouchableOpacity>
    </View>
  );
};

const styles = StyleSheet.create({
  view: {
  },
  inputView: {
    borderColor: '#000',
    borderWidth: 2,
    flex: 3,
    height: '100%',
    padding: '1%',
    marginLeft: '2%',
    marginRight: '2%',
  },
  input: {
    flex: 3,
    height: '100%',
    padding: '1%',
    marginLeft: '2%',
    marginRight: '2%',
    fontSize: 12,
    color: '#777776',
  },
  textView: {
    padding: '1%',
    borderColor: 'yellow',
    borderWidth: 3,
    flex: 3,
  },
  text: {
    fontSize: 17,
    textAlign: 'center',
    justifyContent: 'center',
    alignContent: 'center',
    fontWeight: '700',
  },
  button: {
    backgroundColor: '#62bfac',
    flex: 2.5,
  },
  buttonText: {
    paddingTop: '12%',
    paddingLeft: '20%',
    paddingRight: '20%',
    color: '#fff',
    textAlign: 'center',
  },
});

export default InputForm;
