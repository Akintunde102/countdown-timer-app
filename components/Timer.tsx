import React, {useState, useEffect, useContext} from 'react';
import {
  StyleSheet,
  View,
  Text,
  TouchableOpacity,
  ViewStyle,
} from 'react-native';
import Icon from 'react-native-vector-icons/MaterialIcons';
import {Player} from '@react-native-community/audio-toolkit';
import {useInterval, FormatTimeForDisplay} from '../utils';
import ControlButtons from './ControlButtons';
import {SizeContext} from '../contexts';

const Timer = ({
  durationInMin,
  startDetails,
  style,
}: {
  durationInMin: number;
  startDetails: {starts: boolean; time?: Date};
  style: ViewStyle;
}) => {
  // Convert Duration To Seconds
  const durationInSecs = durationInMin * 60;

  // Contexts
  const {fontScale, dHeight, dWidth} = useContext(SizeContext);

  // States
  const [secsLeft, setSecsLeft] = useState<number>(durationInSecs);
  const [delay, setDelay] = useState<number>(1000);
  const {formattedDisplay, digitNumber} = FormatTimeForDisplay(secsLeft);
  const [pause, setPause] = useState<boolean>(true);
  const [showTimer, setShowTimer] = useState<boolean>(false);
  const [tagText, setTagText] = useState<string>('');
  const {starts: shouldTimerStart, time: initializationTime} = startDetails;

  // To Restart Timer If User Clicks Play Again
  useEffect(() => {
    if (shouldTimerStart) {
      setSecsLeft(durationInSecs);
      setTagText('');
      if (pause) {
        setPause(false);
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [durationInSecs, initializationTime]);

  // An Intelligent Wrapper for UseEffect and SetInterval, controls the timer ticks
  // and watches for operations tide to the Timer ticks
  useInterval(
    () => {
      if (shouldTimerStart) {
        if (!pause && secsLeft <= durationInSecs / 2 + 1) {
          setTagText('More than halfway there!');
        }

        if (secsLeft <= 1) {
          setTagText("Time's Up");
          setPause(true);
        }

        setSecsLeft(secsLeft - 1);
      }
    },
    pause ? null : delay,
  );

  // This Watches for only for Blink And Sound Effects
  // 'cause blink and sound effects needs their own 'smart" interval delays
  useInterval(() => {
    if (!pause && secsLeft <= 10 && secsLeft !== 0) {
      setShowTimer(!showTimer);

      new Player('ping.mp3').play().on('ended', () => {
        console.log('Sound Played');
      });
    }

    if (secsLeft === 0) {
      setShowTimer(true);
    }
  }, delay / 2);

  const styles = StyleSheet.create({
    wholeView: {
      flexDirection: 'column',
      flex: 1,
      backgroundColor: '#ffc',
    },
    mainView: {
      flex: 1,
      flexDirection: 'column',
      marginTop: dHeight * 0.2,
      marginBottom: dHeight * 0.01,
      marginLeft: dWidth * 0.02,
      marginRight: dWidth * 0.02,
    },
    tagTextView: {
      flex: 0.1,
      textAlign: 'center',
      justifyContent: 'center',
      alignItems: 'center',
      marginRight: dWidth * 0.1,
    },
    tagText: {
      fontSize: 20 / fontScale,
      color: '#000',
      fontWeight: '700',
    },
    mainViewTimer: {
      flex: 0.4,
      flexDirection: 'row',
    },
    mainViewText: {
      flex: 6,
      textAlign: 'center',
      justifyContent: 'center',
      alignItems: 'center',
    },
    mainViewIcon: {
      flex: 1,
      textAlign: 'center',
      justifyContent: 'center',
      alignItems: 'center',
      marginBottom: dWidth * 0.05,
      padding: 0,
    },
    timerText: {
      fontFamily: 'uFont',
      textAlign: 'center',
      fontSize: 580 / digitNumber / fontScale,
      ...(shouldTimerStart && secsLeft <= 20 && secsLeft !== 0
        ? {
            color: 'red',
          }
        : {
            color: '#000',
          }),
      ...(showTimer
        ? {}
        : {
            display: 'none',
          }),
    },
    icon: {},
    controlButtons: {
      flex: 0.2,
      flexDirection: 'row',
      borderTopWidth: 10,
      borderRadius: 2,
      borderColor: '#ddd',
      justifyContent: 'center',
    },
  });

  return (
    <View style={{...styles.wholeView, ...style}}>
      <View style={styles.mainView}>
        <View style={styles.tagTextView}>
          <Text style={styles.tagText}> {tagText} </Text>
        </View>
        <View style={styles.mainViewTimer}>
          <View style={styles.mainViewText}>
            <Text style={styles.timerText}>{formattedDisplay}</Text>
          </View>
          <View style={styles.mainViewIcon}>
            <TouchableOpacity
              onPress={() => {
                if (secsLeft <= 0) {
                  setTagText('');
                  setSecsLeft(durationInSecs);
                }
                setPause(!pause);
              }}>
              {pause ? (
                <Icon
                  style={styles.icon}
                  name="play-circle-outline"
                  color="#000"
                  size={45 / fontScale}
                />
              ) : (
                <Icon
                  style={styles.icon}
                  name="pause-circle-outline"
                  color="#000"
                  size={45 / fontScale}
                />
              )}
            </TouchableOpacity>
          </View>
        </View>
      </View>
      <ControlButtons
        customStyle={styles.controlButtons}
        delay={delay}
        setDelay={(d: number) => {
          setDelay(d);
        }}
      />
    </View>
  );
};

export default Timer;
