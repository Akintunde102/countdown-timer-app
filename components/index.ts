import Timer from './Timer';
import InputForm from './InputForm';
import ControlButton from './ControlButton';

export {Timer, InputForm, ControlButton};
