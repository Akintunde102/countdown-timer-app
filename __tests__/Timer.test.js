import 'react-native';
import React from 'react';
import {Timer} from '../components';
import renderer from 'react-test-renderer';

jest.mock('react-native-splash-screen', () => {
  return {
    hide: jest.fn(),
    show: jest.fn(),
  };
});

jest.useFakeTimers();

it('should match snapshot', () => {
  const snap = renderer
    .create(
      <Timer
        style={{}}
        startDetails={{starts: true, time: Math.round(Date.now() / 1000)}}
        durationInMin={1}
      />,
    )
    .toJSON();
  expect(snap).toMatchSnapshot();
});
