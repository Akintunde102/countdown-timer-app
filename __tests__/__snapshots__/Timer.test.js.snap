// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#ffc",
      "flex": 1,
      "flexDirection": "column",
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "column",
        "marginBottom": 13.34,
        "marginLeft": 15,
        "marginRight": 15,
        "marginTop": 266.8,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 0.1,
          "justifyContent": "center",
          "marginRight": 75,
          "textAlign": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000",
            "fontSize": 41.66666666666667,
            "fontWeight": "700",
          }
        }
      >
         
        
         
      </Text>
    </View>
    <View
      style={
        Object {
          "flex": 0.4,
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 6,
            "justifyContent": "center",
            "textAlign": "center",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#000",
              "display": "none",
              "fontFamily": "uFont",
              "fontSize": 241.66666666666669,
              "textAlign": "center",
            }
          }
        >
          01:00
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
            "marginBottom": 37.5,
            "padding": 0,
            "textAlign": "center",
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "#000",
                  "fontSize": 93.75,
                },
                Object {},
                Object {
                  "fontFamily": "Material Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            î€¹
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "borderColor": "#ddd",
        "borderRadius": 2,
        "borderTopWidth": 10,
        "flex": 0.2,
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#6c6c6c",
          "borderColor": "#000",
          "borderWidth": 3,
          "justifyContent": "center",
          "margin": 40.019999999999996,
          "marginLeft": 0,
          "marginRight": 22.5,
          "opacity": 1,
          "paddingBottom": 40.019999999999996,
          "paddingLeft": 22.5,
          "paddingRight": 22.5,
          "paddingTop": 40.019999999999996,
          "textAlign": "center",
          "width": 180,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#fff",
            "fontSize": 41.66666666666667,
            "fontWeight": "700",
          }
        }
      >
        1X
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderColor": "#000",
          "borderWidth": 3,
          "justifyContent": "center",
          "margin": 40.019999999999996,
          "marginLeft": 0,
          "marginRight": 22.5,
          "opacity": 1,
          "paddingBottom": 40.019999999999996,
          "paddingLeft": 22.5,
          "paddingRight": 22.5,
          "paddingTop": 40.019999999999996,
          "textAlign": "center",
          "width": 180,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000",
            "fontSize": 41.66666666666667,
            "fontWeight": "700",
          }
        }
      >
        1.5X
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderColor": "#000",
          "borderWidth": 3,
          "justifyContent": "center",
          "margin": 40.019999999999996,
          "marginLeft": 0,
          "marginRight": 22.5,
          "opacity": 1,
          "paddingBottom": 40.019999999999996,
          "paddingLeft": 22.5,
          "paddingRight": 22.5,
          "paddingTop": 40.019999999999996,
          "textAlign": "center",
          "width": 180,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000",
            "fontSize": 41.66666666666667,
            "fontWeight": "700",
          }
        }
      >
        2X
      </Text>
    </View>
  </View>
</View>
`;
